pam/authmodeselection.go

```diff
@@ -0,0 +1,233 @@
+
+		// Select correct line to ensure model is synchronised
+		for i, a := range m.Items() {
+			a := convertTo[authModeItem](a)
+			if a.id != msg.id {
+				continue
+			}
+			m.Select(i)
+		}
+
+		return m, sendEvent(AuthModeSelected{
+			ID: msg.id,
+		})
+	}
+
+	// interaction events
+	if !m.focused {
+		return m, nil
+	}
+	switch msg := msg.(type) {
+	// Key presses
+	case tea.KeyMsg:
+		switch msg.String() {
+		case "enter":
+			item := m.SelectedItem()
+			if item == nil {
+				return m, nil
+			}
+			authMode := convertTo[authModeItem](item)
+			cmd := selectAuthMode(authMode.id)
+			return m, cmd
+		case "1", "2", "3", "4", "5", "6", "7", "8", "9":
+			// This is necessarily an integer, so above
+			choice, _ := strconv.Atoi(msg.String())
+			items := m.Items()
+			if choice > len(items) {
+				return m, nil
+			}
+			item := items[choice-1]
+			authMode := convertTo[authModeItem](item)
+			cmd := selectAuthMode(authMode.id)
+			return m, cmd
+		}
+	}
+
+	var cmd tea.Cmd
+	m.Model, cmd = m.Model.Update(msg)
+	return m, cmd
+}
+
+// Focus focuses this model.
+func (m *authModeSelectionModel) Focus() tea.Cmd {
+	m.focused = true
+	return nil
+}
+
+// Focused returns if this model is focused.
+func (m *authModeSelectionModel) Focused() bool {
+	return m.focused
+}
+
+// Blur releases the focus from this model.
+func (m *authModeSelectionModel) Blur() {
+	m.focused = false
+}
+
+// WillCaptureEscape returns if this broker may capture Esc typing on keyboard.
+func (m authModeSelectionModel) WillCaptureEscape() bool {
+	return m.FilterState() == list.Filtering
+}
+
+// authModeItem is the list item corresponding to an authentication mode.
+type authModeItem struct {
+	id    string
+	label string
+}
+
+// FilterValue allows filtering the list items.
+func (i authModeItem) FilterValue() string { return "" }
+
+// validAuthModeID returns if a authmode ID exists in the available list.
+func validAuthModeID(id string, authModes []*authd.GAMResponse_AuthenticationMode) bool {
+	for _, a := range authModes {
+		if a.Id != id {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+// getAuthenticationModes returns available authentication mode for this broker from authd.
+func getAuthenticationModes(client authd.PAMClient, sessionID string) tea.Cmd {
+	return func() tea.Msg {
+		required, optional := "required", "optional"
+		supportedEntries := "optional:chars,chars_password"
+		requiredWithBooleans := "required:true,false"
+		optionalWithBooleans := "optional:true,false"
+
+		gamReq := &authd.GAMRequest{
+			SessionId: sessionID,
+			SupportedUiLayouts: []*authd.UILayout{
+				{
+					Type:   "form",
+					Label:  &required,
+					Entry:  &supportedEntries,
+					Wait:   &optionalWithBooleans,
+					Button: &optional,
+				},
+				{
+					Type:    "qrcode",
+					Content: &required,
+					Wait:    &requiredWithBooleans,
+					Label:   &optional,
+				},
+			},
```

Comment:
The `newpassword` one was left out deliberately?

Reply:
yes, that will be built in another PR, once we have the example broker sending this (and chaining MFA)